### PR TITLE
Fix Fork Choice prunning

### PR DESCRIPTION
**Pull request by potuz**
This PR fixes the following issue on current fork choice pruning. Given the following diagram of nodes, where the numbers are the indices in the Fork Choice

```
0 -- 1
  \ _ 2
```
And we finalize the block 1, we should be left with only that block, pruning away the older final state `0` and the contentious fork `2`. Currently we keep a dangling node `2` that gets pruned later. 

WARNING: the loop is very fast, but it is a loop over the whole fork choice, this may not be so critical since it's only applied when we are finalizing a node, therefore either we are in a good case, or if we were forked, we are getting away from trouble.

### CODE REVIEW

**Review by terencechain — CHANGES_REQUESTED** [review]
one question on deleting old indices from map

File: beacon-chain/forkchoice/protoarray/store.go [hunk 1]
```diff
@@ -578,52 +578,42 @@ func (s *Store) prune(ctx context.Context, finalizedRoot [32]byte) error {
...
+	// to fork choice before the finalized block is either final now or
+	// it is in a contentious fork.
+	processedNodes := make(map[uint64]uint64, uint64(len(s.nodes))-finalizedIndex)
+	newNodes := make([]*Node, 1, uint64(len(s.nodes))-finalizedIndex)
+	node := s.nodes[finalizedIndex]
+	node.parent = NonExistentNode
+	newNodes[0] = node
+	processedNodes[finalizedIndex] = uint64(0)
+	for idx := uint64(0); idx < uint64(len(s.nodes)); idx++ {
+		node = s.nodes[idx]
```
Comment (terencechain) [inline beacon-chain/forkchoice/protoarray/store.go:591]: I think no copy is fine here but do we feel safer using a copied node?

File: beacon-chain/forkchoice/protoarray/store.go [hunk 2]
```diff
@@ -578,52 +578,42 @@ func (s *Store) prune(ctx context.Context, finalizedRoot [32]byte) error {
...
+			// This node is a descendant of the finalized point
+			s.nodesIndices[node.root] = uint64(len(newNodes))
+			processedNodes[idx] = uint64(len(newNodes))
+			node.parent = parentIdx
+			newNodes = append(newNodes, node)
+		} else {
+			// This node is in a contentious fork
+			delete(s.nodesIndices, node.root)
 		}
-		delete(s.nodesIndices, s.nodes[i].root)
```
Comment (terencechain) [inline beacon-chain/forkchoice/protoarray/store.go:587]: I think we still need to clean  `s.nodesIndices` where `i < finalizedIndex`

Comment (potuz) [reply beacon-chain/forkchoice/protoarray/store.go:587]: Nodes with `i < finalizedIndex` cannot be descendant of the finalized node, therefore they are deleted in the previous loop. I can update the comment "This node is in a contentious fork" to "This node is not a descentant of the finalized state"

Comment (terencechain) [reply beacon-chain/forkchoice/protoarray/store.go:587]: Ah, i see it. Neat!

**Review by terencechain — DISMISSED** [review]

**Review by terencechain — APPROVED** [review]

**Pull request merged**